cdap-app-fabric/src/test/java/co/cask/cdap/internal/app/store/AppMetadataStoreTest.java

```diff
@@ -0,0 +1,120 @@
+import co.cask.cdap.common.app.RunIds;
+import co.cask.cdap.data2.dataset2.DatasetFramework;
+import co.cask.cdap.internal.AppFabricTestHelper;
+import co.cask.cdap.proto.Id;
+import co.cask.cdap.proto.ProgramRunStatus;
+import co.cask.cdap.proto.ProgramType;
+import com.google.common.base.Function;
+import com.google.common.collect.ImmutableMap;
+import com.google.common.collect.ImmutableSet;
+import com.google.common.collect.Iterables;
+import com.google.inject.Injector;
+import org.apache.twill.api.RunId;
+import org.junit.Assert;
+import org.junit.BeforeClass;
+import org.junit.Test;
+
+import java.util.Set;
+import java.util.TreeSet;
+import java.util.concurrent.TimeUnit;
+
+/**
+ * Test AppMetadataStore.
+ */
+public class AppMetadataStoreTest {
+  private static DatasetFramework datasetFramework;
+  private static Id.Namespace defaultNamespace = Id.Namespace.from("default");
+
+  @BeforeClass
+  public static void beforeClass() throws Exception {
+    Injector injector = AppFabricTestHelper.getInjector();
+    AppFabricTestHelper.ensureNamespaceExists(defaultNamespace);
+    datasetFramework = injector.getInstance(DatasetFramework.class);
+  }
+
+  @Test
+  public void testScanRunningInRangeWithBatch() throws Exception {
+    Id.DatasetInstance storeTable = Id.DatasetInstance.from(defaultNamespace, "testScanRunningInRange");
+    datasetFramework.addInstance(Table.class.getName(), storeTable, DatasetProperties.EMPTY);
+
+    Table table = datasetFramework.getDataset(storeTable, ImmutableMap.<String, String>of(), null);
+    Assert.assertNotNull(table);
+    AppMetadataStore metadataStoreDataset = new AppMetadataStore(table);
+
+    // Add some run records
+    TreeSet<Long> expected = new TreeSet<>();
+    for (int i = 0; i < 100; ++i) {
+      Id.Application application = Id.Application.from(defaultNamespace, "app" + i);
+      Id.Program program = Id.Program.from(application, ProgramType.values()[i % ProgramType.values().length],
+                                           "program" + i);
+      RunId runId = RunIds.generate((i + 1) * 10000);
+      expected.add(RunIds.getTime(runId, TimeUnit.MILLISECONDS));
+      // Start the program and stop it
+      metadataStoreDataset.recordProgramStart(program, runId.getId(), RunIds.getTime(runId, TimeUnit.SECONDS),
+                                              null, null);
+      metadataStoreDataset.recordProgramStop(program, runId.getId(), RunIds.getTime(runId, TimeUnit.SECONDS),
+                                             ProgramRunStatus.values()[i % ProgramRunStatus.values().length]);
+    }
+
+    // Run full scan
+    runScan(metadataStoreDataset, expected, 0, Long.MAX_VALUE);
+
+    // Run the scan with time limit
+    runScan(metadataStoreDataset, expected.subSet(30 * 10000L, 90 * 10000L),
+            TimeUnit.MILLISECONDS.toSeconds(30 * 10000), TimeUnit.MILLISECONDS.toSeconds(90 * 10000));
+
+    runScan(metadataStoreDataset, expected.subSet(90 * 10000L, 101 * 10000L),
```

Comment:
This test is just testing another limit boundary. End time is exclusive for both Treeset and metadataStore.